src/v/s3/client.cc

```diff
@@ -316,6 +311,102 @@ request_creator::make_delete_object_request(
     return header;
 }
 
+struct delete_objects_body : public ss::data_source_impl {
+    ss::temporary_buffer<char> data;
+    explicit delete_objects_body(std::string_view body) noexcept
+      : data{body.data(), body.size()} {}
+    auto get() -> ss::future<ss::temporary_buffer<char>> override {
+        return ss::make_ready_future<ss::temporary_buffer<char>>(
+          std::exchange(data, {}));
+    }
+};
+
+result<std::tuple<http::client::request_header, ss::input_stream<char>>>
+request_creator::make_delete_objects_request(
+  const s3::bucket_name& name, std::span<const object_key> keys) {
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
+    // will generate this request:
+    //
+    // POST /?delete HTTP/1.1
+    // Host: <Bucket>.s3.amazonaws.com
+    // Content-MD5: <Computer from body>
+    // Authorization: <applied by _requestor>
+    // Content-Length: <...>
+    //
+    // <?xml version="1.0" encoding="UTF-8"?>
+    // <Delete>
+    //     <Object>
+    //         <Key>object_key</Key>
+    //     </Object>
+    //     <Object>
+    //         <Key>object_key</Key>
+    //     </Object>
+    //      ...
+    //     <Quiet>true</Quiet>
+    // </Delete>
+    //
+    // note:
+    //  - Delete.Quiet true will generate a response that reports only failures
+    //  to delete or errors
+    //  - the actual xml might not be pretty-printed
+    //  - with clang15 and ranges, xml generation could be a one-liner + a
+    //  custom formatter for xml escaping
+
+    auto body = [&] {
+        auto delete_tree = boost::property_tree::ptree{};
+        // request a quiet response
+        delete_tree.put("Delete.Quiet", true);
+        // add an array of Object.Key=key to the Delete root
+        for (auto key_tree = boost::property_tree::ptree{};
+             auto const& k : keys) {
+            key_tree.put("Key", k().c_str());
+            delete_tree.add_child("Delete.Object", key_tree);
+        }
+
+        auto out = std::ostringstream{};
+        // to indent with 1 tab, pass
+        // boost::property_tree::xml_writer_settings<std::string>('\t', 1) as
+        // third parameter
+        boost::property_tree::write_xml(out, delete_tree);
+        return out.str();
+    }();
```

Comment:
question: what is the use of creating these unnamed lambdas and then calling them right away? 

Is it to restrict scope of the local variables inside them?

Reply:
yes basically.
In this case nothing is really gained over a simple `{ }` block, but with a lambda you gain the use of `return` and early exit, so it can make variable initialization nicer.
for me is just muscle memory